trl/trainer/ppov2_bandit_rloo_trainer.py

```diff
@@ -0,0 +1,704 @@
+            drop_last=True,  # needed; otherwise the last batch will be of ragged shape
+        )
+
+        device = accelerator.device
+        torch.backends.cudnn.deterministic = True # TODO: check if cudnn matters
+
+        # sync random states for DataLoader(shuffle=True) before `accelerator.prepare`
+        # see https://gist.github.com/vwxyzjn/2581bff1e48e185e0b85b6dfe1def79c
+        torch.manual_seed(args.seed)
+        model, optimizer, dataloader = accelerator.prepare(model, optimizer, dataloader)
+        torch.manual_seed(local_seed)  # reset the local seed again
+
+        def repeat_generator():
+            while True:
+                yield from dataloader
+
+        iter_dataloader = iter(repeat_generator())
+        if args.deepspeed2:
+            reward_model = prepare_deepspeed2(reward_model, args.per_device_train_batch_size)
+            ref_policy = prepare_deepspeed2(ref_policy, args.per_device_train_batch_size)
+        elif args.deepspeed3:
+            reward_model = prepare_deepspeed3(reward_model, accelerator)
+            ref_policy = prepare_deepspeed3(ref_policy, accelerator)
+        else:
+            ref_policy = ref_policy.to(device)
+            reward_model = reward_model.to(device)
+
+        generation_config = GenerationConfig(
+            max_new_tokens=args.response_length,
+            min_new_tokens=args.response_length,
+            temperature=(args.temperature + 1e-7),
+            top_k=0.0,
+            top_p=1.0,
+            do_sample=True,
+        )
+
+        accelerator.print("===training policy===")
+        global_step = 0
+        start_time = time.time()
+        stats_shape = (args.noptepochs, args.nminibatches, args.gradient_accumulation_steps)
+        approxkl_stats = torch.zeros(stats_shape, device=device)
+        pg_clipfrac_stats = torch.zeros(stats_shape, device=device)
+        pg_loss_stats = torch.zeros(stats_shape, device=device)
+        vf_loss_stats = torch.zeros(stats_shape, device=device)
+        vf_clipfrac_stats = torch.zeros(stats_shape, device=device)
+        entropy_stats = torch.zeros(stats_shape, device=device)
+        ratio_stats = torch.zeros(stats_shape, device=device)
+        model.train()
+        for _ in range(1, args.num_updates + 1):
+            global_step += 1 * args.batch_size
+            self.lr_scheduler.step()
+            data = next(iter_dataloader)
+            with torch.no_grad():
+                queries = data["input_ids"].to(device)
+                queries = queries.repeat(args.rloo_k, 1)
+                context_length = queries.shape[1]
+                query_responses = []
+                responses = []
+                postprocessed_responses = []
+                logprobs = []
+                ref_logprobs = []
+                scores = []
+                sequence_lengths = []
+                for i in range(0, queries.shape[0], args.local_rollout_forward_batch_size):
```

Comment:
Here you could use the `unwrap_model_for_generation()` context manager as well

Reply:
Will check it out